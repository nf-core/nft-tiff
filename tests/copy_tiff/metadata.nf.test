nextflow_process {

    name "Validate tiff metadata"
    script "./main.nf"
    process "COPY_TIFF"

    tag "metadata"

    test("check metadata equivalence") {
        when {
            process {
                """
                input[0] = Channel.of(
                    file("${baseDir}/tests/mindagap.mouse_heart.wga.tiff", checkIfExists:true),
                )
                """
            }
        }
        then {
            def tiff = path(process.out.tiff[0]).tiff
            def reference = path("${baseDir}/tests/mindagap.mouse_heart.wga.tiff").tiff
            assert tiff.meta == reference.meta
        }
    }

    test("check tiff dimension equivalence") {
        when {
            process {
                """
                input[0] = Channel.of(
                    file("${baseDir}/tests/nuclear_image.tif", checkIfExists:true),
                )
                """
            }
        }
        then {
            def tiff = path(process.out.tiff[0]).tiff
            def dimensions = tiff.meta.getDimensions()

            assert snapshot(dimensions).match()
        }
    }

    test("check metadata field count and field names equivalence") {
        when {
            process {
                """
                input[0] = Channel.of(
                    file("${baseDir}/tests/mindagap.mouse_heart.wga.tiff", checkIfExists:true),
                )
                """
            }
        }
        then {
            def tiff = path(process.out.tiff[0]).tiff
            def fieldCount = tiff.meta.getFieldCounts()
            def fieldNames = tiff.meta.getFieldNames()

            assert snapshot(
                fieldCount,
                fieldNames
            ).match()
        }
    }

    test("check metadata value equivalence for selected fields") {
        when {
            process {
                """
                input[0] = Channel.of(
                    file("${baseDir}/tests/mindagap.mouse_heart.wga.tiff", checkIfExists:true),
                )
                """
            }
        }
        then {
            def tiff = path(process.out.tiff[0]).tiff
            def fieldNames = ['ImageWidth', 'PlanarConfiguration', 'Compression']
            def fields = tiff.meta.getFields(fieldNames)

            assert snapshot(fields).match()
        }
    }

    test("check metadata value equivalence for single field") {
        when {
            process {
                """
                input[0] = Channel.of(
                    file("${baseDir}/tests/mindagap.mouse_heart.wga.tiff", checkIfExists:true),
                )
                """
            }
        }
        then {
            def tiff = path(process.out.tiff[0]).tiff
            def fieldName = 'ImageWidth'
            def fields = tiff.meta.getFields(fieldName)

            assert snapshot(fields).match()
        }
    }

    test("check metadata value equivalence for all fields") {
        when {
            process {
                """
                input[0] = Channel.of(
                    file("${baseDir}/tests/mindagap.mouse_heart.wga.tiff", checkIfExists:true),
                )
                """
            }
        }
        then {
            def tiff = path(process.out.tiff[0]).tiff
            def fields = tiff.meta.getFields()

            assert snapshot(fields).match()
        }
    }
    


    test("detect invalid metadata field") {
        when {
            process {
                """
                input[0] = Channel.of(
                    file("${baseDir}/tests/mindagap.mouse_heart.wga.tiff", checkIfExists:true),
                )
                """
            }
        }
        then {
            def tiff = path(process.out.tiff[0]).tiff
            def fieldName = 'NotValid'

            try {
                tiff.meta.getFields(fieldName)
            } catch (Exception e)  {
                assert e.message == "Invalid field tag: NotValid"
            }
        }
    }
}
