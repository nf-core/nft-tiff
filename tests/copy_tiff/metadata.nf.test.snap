{
    "check metadata value equivalence for single field": {
        "content": [
            [
                {
                    "ImageWidth": "528"
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-02T14:44:57.573898328"
    },
    "check tiff dimension equivalence": {
        "content": [
            [
                520,
                696
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-02T14:44:43.94436687"
    },
    "check metadata value equivalence for selected fields": {
        "content": [
            [
                {
                    "PlanarConfiguration": null,
                    "Compression": "1",
                    "ImageWidth": "528"
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-02T14:44:53.288343043"
    },
    "check metadata field count and field names equivalence": {
        "content": [
            [
                11
            ],
            [
                [
                    "NewSubfileType",
                    "ImageWidth",
                    "ImageLength",
                    "BitsPerSample",
                    "Compression",
                    "PhotometricInterpretation",
                    "ImageDescription",
                    "StripOffsets",
                    "SamplesPerPixel",
                    "RowsPerStrip",
                    "StripByteCounts"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-02T14:44:48.919742217"
    },
    "check metadata value equivalence for all fields": {
        "content": [
            [
                {
                    "RowsPerStrip": "528",
                    "Compression": "1",
                    "PhotometricInterpretation": "1",
                    "SamplesPerPixel": "1",
                    "ImageWidth": "528",
                    "StripByteCounts": "[557568]",
                    "StripOffsets": "[180]",
                    "ImageLength": "528",
                    "ImageDescription": "[ImageJ=1.53t\nmin=0.0\nmax=16383.0\n]",
                    "NewSubfileType": "0",
                    "BitsPerSample": "[16]"
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-02T14:45:02.756392834"
    }
}